### PR TITLE
refactor message module

### LINKED ISSUE #799: Refactor Message module
### Is your feature request related to a problem?

Actually all serialization and deserialization are done in the message module, but as the number of message growth, the reaadability of this module and it's testing begin to be complex.

### Describe the solution you'd like

We should refactor this module to improve test and readability.
The serialization and deserialization could be done in the module of the message itself, an actual example is the `ShardRepair` message which use
```elixir
def encode(msg = %ShardRepair{}) do
  <<230::8, ShardRepair.serialize(msg)::bitstring>>
end
```
The testing part of serialization and deserialization can be done in doctest of this message

We can also put in the message itself the process, for example: 
```elixir
def process(msg = %ShardRepair{}) do
  ShardRepair.process(msg)
end
```
To go further we don't need to patern match de message type, but directly call the structure, so there would be only 1 function process in the Message module:
```elixir
def process(msg) do
  msg.__struct__.process(msg)
end
```

To encode and decode a message we need to know the id number of it. We could create a correspondence table like 
```elixir
id_map = %{
  ShardRepair => <<230>>,
  ... => ...
}
```
So we could have 1 function for encoding and decoding :
```elixir
def encode(msg) do
  msg_id = Map.get(id_map, msg.__struct__)
  serialized = msg.__struct__.serialize(msg)
  <<msg_id::8, serialized::binary>>
end

def decode(<<msg_id::8, rest::binary>>) do
  {msg_type, _} = Enum.find(id_map, fn {key, value} -> value == msg_id end)
  msg_type.deserialize(rest)
end
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tenmoves**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #799 

## Type of change

Please delete options that are not relevant.

-  Refactoring

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by bchamagne — COMMENTED** [review]
Suggestion:

The encode function is not doing much, it is only doing work when there is no serialize/1 function. 

Is it possible to keep the encode/decode functions in the message module as one liners per clause? This way we could keep all the prefix in the same spot?

```
@spec encode(t()) :: bitstring()
def encode(msg = %GetBoostrappingNodes{}), do: <<0::8, msg.__struct__.serialize(msg)
def encode(msg = %GetStorageNonce{}), do: <<1::8, msg.__struct__.serialize(msg)
...

@spec decode(bitstring) :: t()
def decode(<<0::8, rest::bitstring), do: GetBootstrappingNodes.deserialize(rest)
def decode(<<1::8, rest::bitstring), do: GetStorageNonce.deserialize(rest)
...
```

File: lib/archethic/transaction_chain/transaction.ex [hunk 1]
```diff
@@ -900,4 +900,9 @@ defmodule Archethic.TransactionChain.Transaction do
         |> Enum.map(&CrossValidationStamp.cast/1)
     }
   end
+
+  @spec encode(t()) :: bitstring()
```
Comment (bchamagne) [inline lib/archethic/transaction_chain/transaction.ex:904]: can be removed?

File: lib/archethic/p2p/message.ex [hunk 2]
```diff
@@ -237,1731 +208,26 @@ defmodule Archethic.P2P.Message do
...
-  end
-
-  def decode(<<254::8, rest::bitstring>>) do
-    {%Ok{}, rest}
+  def decode(<<message_id::8, rest::bitstring>>) do
+    msg_module = MessageId.id_to_module(message_id)
+    msg_module.deserialize(rest)
   end
 
   def decode(<<255::8>>), do: raise("255 message type is reserved for stream EOF")
```
Comment (bchamagne) [inline lib/archethic/p2p/message.ex:226]: This clause should be before the precedent:

```
iex(1)> Archethic.P2P.Message.decode(<<255::8>>)
** (FunctionClauseError) no function clause matching in Archethic.P2P.MessageId.id_to_module/1   
```

**Review by bchamagne — APPROVED** [review]
minor comments, looks great!

File: lib/archethic/p2p/message.ex (hunk 2 above)
Comment (tenmoves) [reply lib/archethic/p2p/message.ex:226]: nice catch

**Comment by Neylix** [comment]
For the conflict, there is actually 2 changes:
- GetFirstAddress and FirstAddress message has been renamed (including file name)
- 3 new messages is added

**Review by tenmoves — COMMENTED** [review]
excellent the metaprogramming part

**Review by Neylix — CHANGES_REQUESTED** [review]

**Review by Neylix — APPROVED** [review]

**Pull request merged**